File: glass-xe/src/main/java/com/damn/anotherglass/glass/host/core/BatteryStatus.java
package com.damn.anotherglass.glass.host.core;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.BatteryManager;

import com.damn.anotherglass.shared.device.BatteryStatusData;

public class BatteryStatus {

    public interface Listener {
        void onBatteryStatusChanged(BatteryStatusData data);
    }

    private final Context mContext;
    private final Listener mListener;
    private BroadcastReceiver mReceiver;

    public BatteryStatus(Context context, Listener listener) {
        mContext = context;
        mListener = listener;
    }

    public void start() {
        if (mReceiver != null) {
            return; // already started
        }
        mReceiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                if (intent == null) return;
                int level = intent.getIntExtra(BatteryManager.EXTRA_LEVEL, -1);
                int scale = intent.getIntExtra(BatteryManager.EXTRA_SCALE, -1);
                int percentage = -1;
                if(level >= 0 && scale > 0) {
                    percentage = (int) ((level * 100.0) / scale);
                }
                int status = intent.getIntExtra(BatteryManager.EXTRA_STATUS, -1);
                boolean isCharging = status == BatteryManager.BATTERY_STATUS_CHARGING ||
                                     status == BatteryManager.BATTERY_STATUS_FULL;

                if (mListener != null) {
                    mListener.onBatteryStatusChanged(new BatteryStatusData(percentage, isCharging));
                }
            }
        };

        IntentFilter filter = new IntentFilter(Intent.ACTION_BATTERY_CHANGED);
        mContext.registerReceiver(mReceiver, filter);

        Intent stickyIntent = mContext.registerReceiver(null, filter);
        if(stickyIntent != null) {
            mReceiver.onReceive(mContext, stickyIntent);
        }
    }

    public void stop() {
        if (mReceiver != null) {
            mContext.unregisterReceiver(mReceiver);
            mReceiver = null;
        }
    }
}

File: glass-ee/src/main/java/com/damn/anotherglass/glass/ee/host/core/BatteryStatus.kt
package com.damn.anotherglass.glass.ee.host.core

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.BatteryManager
import androidx.lifecycle.MutableLiveData

class BatteryStatus(private val context: Context) : MutableLiveData<BatteryStatus.BatteryStatusData>() {

    data class BatteryStatusData(val level: Int, val isCharging: Boolean)

    private val receiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context?, intent: Intent?) {
            if (intent == null) return
            val level = intent.getIntExtra(BatteryManager.EXTRA_LEVEL, -1)
            val scale = intent.getIntExtra(BatteryManager.EXTRA_SCALE, -1)
            val percentage = when {
                level >= 0 && scale > 0 -> (level * 100.0 / scale).toInt()
                else -> -1
            }
            val status = intent.getIntExtra(BatteryManager.EXTRA_STATUS, -1)
            val isCharging = BatteryManager.BATTERY_STATUS_CHARGING == status ||
                    BatteryManager.BATTERY_STATUS_FULL == status

            postValue(BatteryStatusData(percentage, isCharging))
        }
    }

    override fun onActive() {
        super.onActive()
        context.registerReceiver(receiver, IntentFilter(Intent.ACTION_BATTERY_CHANGED))
        val batteryStatus = IntentFilter(Intent.ACTION_BATTERY_CHANGED).let {
            context.registerReceiver(null, it)
        }
        receiver.onReceive(context, batteryStatus)
    }

    override fun onInactive() {
        super.onInactive()
        context.unregisterReceiver(receiver)
    }

    companion object {
        @JvmStatic
        fun batteryStatusString(data: BatteryStatusData): String {
            // todo: use drawables for battery icon
            val icon = "\uD83D\uDD0B" // glass does not support low battery icon
            return "$icon ${data.level}%${if (data.isCharging) "⚡" else ""}"
        }
    }
}